File: c++ version/1-100/46-permutations.cpp
vector<vector<int>> permute(vector<int>& nums) {
    vector<vector<int>> ans;
    backtracking(nums, 0, ans);
    return ans;
}

void backtracking(vector<int> &nums, int pos, vector<vector<int>> &ans) {
    if (pos == nums.size() - 1) {
        ans.push_back(nums);
        return;
    }
    for (int i = pos; i < nums.size(); ++i) {
        swap(nums[i], nums[pos]);
        backtracking(nums, pos + 1, ans);
        swap(nums[i], nums[pos]);
    }
}

File: python version/46-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        visited = [False for _ in range(n)]
        curr = []
        ans = []

        def backtracking():
            if len(curr) == n:
                ans.append(curr[:])
                return
            for i in range(n):
                if visited[i] == False:
                    visited[i] = True
                    curr.append(nums[i])
                    backtracking()
                    curr.pop()
                    visited[i] = False
        
        backtracking()
        return ans
